### PR TITLE
run apply on multiple stacks in porter yaml

### PR DESCRIPTION
**Pull request by ianedwards**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

### CODE REVIEW

File: cli/cmd/stack/apply.go [hunk 1]
```diff
@@ -3,98 +3,159 @@ package stack
...
-	v1File := &switchboardTypes.ResourceGroup{
-		Version: "v1",
-		Resources: []*switchboardTypes.Resource{
-			{
-				Name:   "get-env",
-				Driver: "os-env",
-			},
-		},
-	}
+func CreateStackDeploy(client *api.Client, worker *switchboardWorker.Worker, app *Application, stackName string, cliConf *config.CLIConfig) ([]*switchboardTypes.Resource, error) {
```
Comment (ferozemohideen) [inline cli/cmd/stack/apply.go:26]: should we call this createApplicationDeploy?

Comment (ferozemohideen) [reply cli/cmd/stack/apply.go:26]: also change the parameter stackName to applicationName or just name

File: cli/cmd/stack/apply.go [hunk 2]
```diff
@@ -3,98 +3,159 @@ package stack
...
+	// we need to know the builder so that we can inject launcher to the start command later if heroku builder is used
 	var builder string
+	resources, builder, err := createV1BuildResources(client, app, stackName, cliConf.Project, cliConf.Cluster)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing porter.yaml for build resources: %w", err)
+	}
 
-	stackConf, err := createStackConf(client, raw, stackName, projectID, clusterID)
+	stackBytes, err := yaml.Marshal(app)
```
Comment (ferozemohideen) [inline cli/cmd/stack/apply.go:34]: applicationBytes

File: cli/cmd/stack/apply.go [hunk 3]
```diff
@@ -3,98 +3,159 @@ package stack
...
+		StackName:            stackName,
+		ProjectID:            cliConf.Project,
+		ClusterID:            cliConf.Cluster,
+		BuildImageDriverName: GetBuildImageDriverName(stackName),
+		PorterYAML:           stackBytes,
+		Builder:              builder,
+	}
+	worker.RegisterHook("deploy-stack", deployStackHook)
+
+	if os.Getenv("GITHUB_RUN_ID") != "" {
```
Comment (ferozemohideen) [inline cli/cmd/stack/apply.go:50]: probably best to move all the create event stuff to a helper method

File: cli/cmd/stack/apply.go [hunk 4]
```diff
@@ -3,98 +3,159 @@ package stack
...
-		pi.Name,
-		stackConf.projectID,
-		stackConf.clusterID,
-		stackConf.parsed.Env,
-	)
+func createV1BuildResources(client *api.Client, app *Application, stackName string, projectID uint, clusterID uint) ([]*switchboardTypes.Resource, string, error) {
+	var builder string
+	resources := make([]*switchboardTypes.Resource, 0)
+
+	stackConf, err := createStackConf(client, app, stackName, projectID, clusterID)
```
Comment (ferozemohideen) [inline cli/cmd/stack/apply.go:100]: let's rename all stack stuff to application for clarity. Might want to rename this entire folder as well

File: cli/cmd/stack/build.go [hunk 5]
```diff
@@ -141,14 +141,14 @@ func (b *Build) getV1PushImage(namespace string) (*types.Resource, error) {
 	}
 
 	return &types.Resource{
-		Name:   b.GetName(),
+		Name:   b.GetName(appName),
```
Comment (ferozemohideen) [inline cli/cmd/stack/build.go:144]: can you add "-push-image" to the end of this driver name? to mirror the build driver. not sure if that's gonna break anything tho . but would be clearer in the github action logs what's happening

File: cli/cmd/stack/build.go [hunk 6]
```diff
@@ -120,18 +120,18 @@ func (b *Build) getV1BuildImage(env map[string]string, namespace string) (*types
 	}, nil
 }
 
-func getBuildImageName() string {
-	return "base-image"
+func getBuildImageName(appName string) string {
+	return fmt.Sprintf("%s-base-image", appName)
```
Comment (ferozemohideen) [inline cli/cmd/stack/build.go:124]: do we even need to 'base-image' part anymore? can the build driver name just be appName-build-image?

File: api/server/handlers/porter_app/parse.go [hunk 7]
```diff
@@ -310,21 +280,26 @@ func buildPreDeployJobChartValues(release *App, env map[string]string, synced_en
...
-		return *app.Type
+func getType(name string, service *stack.Service) string {
+	if service.Type != nil {
+		return *service.Type
 	}
 	if strings.Contains(name, "web") {
 		return "web"
 	}
+
+	if strings.Contains(name, "job") {
```
Comment (ferozemohideen) [inline api/server/handlers/porter_app/parse.go:291]: nice

**Review by ferozemohideen — COMMENTED** [review]
Overall looks good to me, most of my suggestions were renaming/moving stuff around. 
I assume you've tested this extensively but for safety I would recommend creating an application locally with and without porter.yaml using the current code, make sure it builds through the GHA, then switching to your new code and running porter apply again. Just to make sure that current applications with/without porter.yamls won't have their github actions fail when this code is pushed

File: cli/cmd/stack/apply.go (hunk 4 above)
Comment (ianedwards) [reply cli/cmd/stack/apply.go:100]: I might make this a follow up task just so I can test that separately if that sounds cool

**Review by ferozemohideen — APPROVED** [review]

**Comment by ferozemohideen** [comment]
https://www.notion.so/porter-run/Github-Workflow-d6b3794d7f264f9f86c925fb08523dac?pvs=4#7cfb26ec4edc400883962083ac5c0485

**Pull request merged**